Spack.md

```diff
@@ -0,0 +1,264 @@
+| o | | | | | | | | |  gdbm
+| o | | | | | | | | |  readline
+````
+
+## Configuring Kokkos as a Project Dependency
+Say you have a project "SuperScience" which needs to use Kokkos.
+In your `package.py` file, you would generally include something like:
+````python
+class SuperScience(CMakePackage):
+  ...
+  depends_on("kokkos")
+````
+Often projects want to tweak behavior when using certain features, e.g.
+````python
+  depends_on("kokkos+cuda", when="+cuda")
+````
+if your project needs CUDA-specific logic to configure and build.
+This illustrates the general principle in Spack of "flowing-up".
+A user requests a feature in the final app:
+````bash
+> spack install superscience+cuda
+````
+This flows upstream to the Kokkos dependency, causing the `kokkos+cuda` variant to build.
+The downstream app (SuperScience) tells the upstream app (Kokkos) how to build.
+
+Because Kokkos is a performance portability library, it somewhat inverts this principle.
+Kokkos "flows-down", telling your application how best to configure for performance.
+Rather than a downstream app (SuperScience) telling the upstream (Kokkos) what variants to build,
+a pre-built Kokkos should be telling the downstream app SuperScience what variants to use.
+Kokkos works best when there is an "expert" configuration installed on your system.
+Your build should simply request `-DKokkos_ROOT=<BEST_KOKKOS_FOR_MY_SYSTEM>` and configure appropriately based on the Kokkos it finds.
+
+Kokkos has many, many build variants.
+Where possible, projects should only depend on a general Kokkos, not specific variants.
+We recommend instead adding for each system you build on a Kokkos configuration to your `packages.yaml` file (usually found in `~/.spack` for specific users).
+For a Xeon + Volta system, this could look like:
+````yaml
+ kokkos:
+  variants: +cuda +openmp +cuda_lambda +wrapper ^cuda@10.1 cuda_arch=70
+  compiler: [gcc@7.2.0]
+````
+which gives the "best" Kokkos configuration as CUDA+OpenMP optimized for a Volta 70 architecture using CUDA 10.1.
+It also enables support for CUDA Lambdas.
+The `+wrapper` option tells Kokkos to build with the special `nvcc_wrapper` (more below).
+Note here that we use the built-in `cuda_arch` variant of Spack to specify the archicture.
+For a Haswell system, we use
+````yaml
+ kokkos:
+  variants: +openmp std=14 target=haswell
+  compiler: [intel@18]
+````
+which uses the built-in microarchitecture variants of Spack.
+Consult the Spack documentation for more details of Spack microarchitectures
+and CUDA architectures.
+Spack does not currently provide an AMD GPU microarchitecture option.
+If building for HIP or an AMD GPU, Kokkos provides an `amd_gpu_arch` similar to `cuda_arch`.
+````yaml
+ kokkos:
+  variants: +hip amd_gpu_arch=vega900
+````
+Note: there is currently no support for the `hipcc` compiler in Spack. Spack currently does not have rules for building the compiler nor can an already installe version be added to Spack as a valid compiler.
```

Comment:
```suggestion
Note: there is currently no support for the `hipcc` compiler in Spack. Spack currently does not have rules for building the compiler nor can an already installed version be added to Spack as a valid compiler.
```
That's not entirely true. You could trick `Spack` into thinking that `hipcc` is a `clang` compiler. That's what I did when testing the `spack` module. I am not sure if we want to mention any of this at all.

Reply:
I just took out that language for now. It's probably not even worth mentioning since hipcc is an advanced use case at this point.